### DISCUSSION TITLE
Versioning and Releases etc.

### DISCUSSION THREAD

**Discussion by adamretter** [eclipse-theia/theia#8445] [General] [answered]
We have been trying to get a release of our Theia IDE based app out this week. We were previously specifying `^1.3.0` in our `package.json` files for Theia dependencies. However, we have API breakage (with `1.4.0`, `1.5.0`). We tried using `~1.3.0` but then we had problems, whereby (for example) `@theia/core` 1.3.0, has a dependency on other theia packages, but uses `^1.3.0` in its own `package.json`, so we saw multiple versions of the same packages being pulled in.

What is the correct way to depend on a fixed released version of Theia please?

**Comment by adamretter** [comment]
From @vince-fugnitto 

>All the @theia dependencies will need to be the same version, and if you require a specific version such as 1.3.0 then it will need to be reflected as such in the package.json. The use of yarn.lock files are also advised for products to pin specific versions of dependencies for which the product requires.
>
> Another approach could be to use yarn resolutions to specifically pin the dependency versions of extensions but I do not believe this is necessary.

**Comment by adamretter** [comment]
@vince-fugnitto It is not clear to me how to pin things in `yarn.lock`. Whenever we do a fresh clone (which includes the `yarn.lock` file), and then a build, we find that yarn has updated our `yarn.lock` file

**Reply by vince-fugnitto** [reply]
@8443 I investigated a bit further, while the `yarn.lock` is useful for applications, it will not pin itself the dependency tree of `@theia` dependencies, the only way to do it at the moment would be to use `resolutions`.

The following comment on a previous discussions may help you further:

- https://github.com/eclipse-theia/theia-cpp-extensions/issues/106#issuecomment-670591239

**Comment by mcgordonite** [comment]
I found that setting a resolutions entry for every @theia extension in the dependency tree guaranteed only one copy of each in the installed node_modules.

**Comment by vince-fugnitto** [comment, marked as answer]
It should be much easier to depend on a specific release now thanks to https://github.com/eclipse-theia/theia/pull/8880 which fixes https://github.com/eclipse-theia/theia/issues/8879.
